### ISSUE TITLE
lsp--text-document-did-close fails for buffers with no file associated

### ISSUE THREAD

**Issue by Fuco1** [emacs-lsp/lsp-mode#207]
# how to reproduce

Invoke magit blame and try to kill the blame buffer.

I fixed it by wrapping the function inside `when buffer-file-name`, not sure what it does really so I don't know if this is appropriate.

There's also related error in `lsp--send-changes` where it tries to `with-current-buffer` on a buffer that was alerady killed... will error with "selecting deleted buffer".  I fixed that by wrapping it with `(buffer-live-p buffer)` check. 

I'm not really sure what is going on but I need to work now so I just hacked it in this way, seems to work fine. Let me know if you need some more info.

**Comment by jaelsasser** [comment]
I've run into the same problem, but `(when buffer-file-name lsp-MODE-enable)` didn't do the trick for me. I'm using this instead:

```elisp
(defun jae--setup-cquery ()
  (when (and buffer-file-name (not magit-buffer-revision))
    (lsp-cquery-enable))
```

**Comment by vibhavp** [comment]
I've pushed a fix at https://github.com/emacs-lsp/lsp-mode/pull/208. Could you try testing it?

**Comment by jaelsasser** [comment]
Went through and tested this without the hack I mentioned above. (I forgot that I was still using it after all these weeks…). Looks like there's still some rough edges (using the latest `cquery`):

* navigate to a version-controlled project that's set up with a `compile_commands.json`
* `lsp-mode` starts up
* check out a revision using `magit-find-file`
* calling `xref-find-defintions` results in the following error message: `url-hexify-string: Wrong type argument: arrayp, nil` (see the attached backtrace)
* quitting the buffer results in the following error message: `Error on ‘lsp--text-document-did-close’: (wrong-type-argument arrayp nil)`

Backtrace for the `xref-find-definitions` error:

```emacs-lisp
Debugger entered--Lisp error: (wrong-type-argument arrayp nil)
  file-truename(nil)
  (url-hexify-string (file-truename buffer-file-name) url-path-allowed-chars)
  (concat lsp--uri-file-prefix (url-hexify-string (file-truename buffer-file-name) url-path-allowed-chars))
  (list :uri (concat lsp--uri-file-prefix (url-hexify-string (file-truename buffer-file-name) url-path-allowed-chars)))
  (or nil (list :uri (concat lsp--uri-file-prefix (url-hexify-string (file-truename buffer-file-name) url-path-allowed-chars))))
  (list :textDocument (or nil (list :uri (concat lsp--uri-file-prefix (url-hexify-string (file-truename buffer-file-name) url-path-allowed-chars)))) :position (or nil (list :line (1- (line-number-at-pos)) :character (- (point) (line-beginning-position)))))
  (propertize (symbol-name (symbol-at-point)) 'def-params (list :textDocument (or nil (list :uri (concat lsp--uri-file-prefix (url-hexify-string (file-truename buffer-file-name) url-path-allowed-chars)))) :position (or nil (list :line (1- (line-number-at-pos)) :character (- (point) (line-beginning-position))))) 'ref-params (lsp--make-reference-params))
  (progn (propertize (symbol-name (symbol-at-point)) 'def-params (list :textDocument (or nil (list :uri (concat lsp--uri-file-prefix (url-hexify-string (file-truename buffer-file-name) url-path-allowed-chars)))) :position (or nil (list :line (1- (line-number-at-pos)) :character (- (point) (line-beginning-position))))) 'ref-params (lsp--make-reference-params)))
  (closure (cl-struct-lsp--workspace-tags cl-struct-lsp--registered-capability-tags cl-struct-lsp--client-tags t) (_backend) (progn (propertize (symbol-name (symbol-at-point)) 'def-params (list :textDocument (or nil (list :uri (concat lsp--uri-file-prefix (url-hexify-string (file-truename buffer-file-name) url-path-allowed-chars)))) :position (or nil (list :line (1- (line-number-at-pos)) :character (- (point) (line-beginning-position))))) 'ref-params (lsp--make-reference-params))))(xref-lsp)
  apply((closure (cl-struct-lsp--workspace-tags cl-struct-lsp--registered-capability-tags cl-struct-lsp--client-tags t) (_backend) (progn (propertize (symbol-name (symbol-at-point)) 'def-params (list :textDocument (or nil (list :uri (concat lsp--uri-file-prefix (url-hexify-string (file-truename buffer-file-name) url-path-allowed-chars)))) :position (or nil (list :line (1- (line-number-at-pos)) :character (- (point) (line-beginning-position))))) 'ref-params (lsp--make-reference-params)))) xref-lsp nil)
  xref-backend-identifier-at-point(xref-lsp)
  xref--read-identifier("Find references of: ")
  byte-code("\300\301!C\207" [xref--read-identifier "Find references of: "] 2)
  call-interactively(xref-find-references nil nil)
  command-execute(xref-find-references)
```

@vibhavp I think `lsp-mode` may still to avoid activating when `buffer-file-name` is `nil`. We handle it a lot cleaner now (before you were forced to hard-quit out of Emacs entirely to recover), but there's still some lingering issues.

**Comment by offby1** [comment]
I'm still occasionally seeing this error, and yet I have the fix from #208.

```
Debugger entered--Lisp error: (wrong-type-argument arrayp nil)
  file-truename(nil)
  lsp--text-document-did-close()
  kill-buffer(#<buffer src/skynet_cli_tools_common/helpers.py.~8ac7082~>)
```
FWIW, the buffer in question, based on its name, looks like it was created by `magit-find-file`.

**Comment by yyoncho** [comment]
Fixed already fixed: d92df3c
